### PR TITLE
deleting trigger and filters created as part of stack deletion

### PR DESCRIPTION
**Pull request by hrai-nr**
After this code changes:
1. On stack deletion, we will delete any S3 trigger event notification created as part of this particular stack.
2. On stack deletion, we will delete any subscription filter created on any log group created as part of this particular stack

Testing:
1. Create a stack, give a S3 bucket name, see that event notification are created. Delete the stack, event notification should also be deleted.
2. Take a S3 bucket, create a sample event trigger for some other event type other than s3:ObjectCreated:*, Now go and create a cloud formation stack here give the same S3 bucket name. Now go and check that there will be 2 event notification on this bucket. Delete the Cloud formation stack. Only the second event notification should get deleted. Any event notification created before stack creation should not get deleted. 
3. Create a stack with selecting log group name, see that a subscription filter is created. Delete the stack, the subscription filter should be deleted.
4. Take a log group with existing subscription filter already present. Create a cloud formation stack with this particular stack, see that second subscription filter get created. Now delete this stack, only the second subscription filter should be deleted. The subscription filter created prior to our stack creation should not get deleted.
5. It is working for prefix as well. Here is how I tested:
- Create a bucket B1. and then create two folders (prefix) inside P1 and P2.
- Go and manually add trigger for P1 and point it to some lambda xyz.
- Now create a stack and give bucket B1 and prefix P2.
- You can see two triggers for the bucket.
- Delete the stack, only the trigger for P2 will be deleted since this will be associated with our lambda.

### CODE REVIEW

File: lambda-cloudwatch-trigger-stack.yaml [hunk 1]
```diff
@@ -116,6 +117,9 @@ Resources:
 
                           add_permission_if_needed(event,context,lambda_arn, log_group_arn, log_group_name)
 
+                          # Unique filter name for this stack using Lambda ARN
+                          filter_name = f'LambdaSubscription_{hashlib.sha256(lambda_arn.encode()).hexdigest()[:20]}'
```
Comment (maya-jha) [inline lambda-cloudwatch-trigger-stack.yaml:121]: maybe add NewRelic as prefix.

File: s3-trigger-stack.yaml [hunk 2]
```diff
@@ -79,9 +79,32 @@ Resources:
...
+                          lambda_arn = event_data.get('LambdaFunctionArn', '')
+                          buckets = event_data.get('S3BucketArns', [])
+                          for bucket_arn in buckets:
+                              parts = bucket_arn.split(":::")[-1].split("/", 1)
+                              bucket_name = parts[0]
+                           
+                              # get the existing notification configuration
+                              existing_config = s3.get_bucket_notification_configuration(Bucket=bucket_name)
+
+                              existing_lambda_configs = existing_config.get('LambdaFunctionConfigurations', [])
```
Comment (maya-jha) [inline s3-trigger-stack.yaml:94]: if this is empty, just continue the loop.

File: s3-trigger-stack.yaml [hunk 3]
```diff
@@ -79,9 +79,32 @@ Resources:
...
+                              updated_lambda_configs = [config for config in existing_lambda_configs if config['LambdaFunctionArn'] != lambda_arn]
```
Comment (maya-jha) [inline s3-trigger-stack.yaml:95]: add comment to reflect that this is effectively deleting the lambda config.

File: s3-trigger-stack.yaml [hunk 4]
```diff
@@ -79,9 +79,32 @@ Resources:
...
+                                  'LambdaFunctionConfigurations': updated_lambda_configs
+                              }
+                              s3.put_bucket_notification_configuration(
+                                  Bucket=bucket_name,
+                                  NotificationConfiguration=notification_config
+                              )
+                      except Exception as e:
+                          logger.error(f'Delete failed for the bucket triggers  with error: {str(e)}')
+                          cfnresponse.send(event, context, cfnresponse.SUCCESS, {})
+                          return {}
```
Comment (maya-jha) [inline s3-trigger-stack.yaml:107]: why return?

File: s3-trigger-stack.yaml [hunk 5]
```diff
@@ -79,9 +79,32 @@ Resources:
               response = {}
               try:
                   if event['RequestType'] == 'Delete':
-                      cfnresponse.send(event, context, cfnresponse.SUCCESS, {})
-                      return {}
-                  
+                      try:
+
```
Comment (maya-jha) [inline s3-trigger-stack.yaml:83]: nit: extra line.

File: s3-trigger-stack.yaml [hunk 6]
```diff
@@ -155,7 +178,7 @@ Resources:
                           .format(','.join(bucket_where_trigger_exists))
                           )
 
-                      cfnresponse.send(event, context, cfnresponse.SUCCESS, response)
+                  cfnresponse.send(event, context, cfnresponse.SUCCESS, response)
```
Comment (maya-jha) [inline s3-trigger-stack.yaml:181]: Why this change?

File: s3-trigger-stack.yaml [hunk 7]
```diff
@@ -79,9 +79,32 @@ Resources:
...
               try:
                   if event['RequestType'] == 'Delete':
-                      cfnresponse.send(event, context, cfnresponse.SUCCESS, {})
-                      return {}
-                  
+                      try:
+
+                          event_data = event['ResourceProperties']
+                          lambda_arn = event_data.get('LambdaFunctionArn', '')
+                          buckets = event_data.get('S3BucketArns', [])
```
Comment (maya-jha) [inline s3-trigger-stack.yaml:86]: you can move it outside if branch as its common for delete and create.

File: s3-trigger-stack.yaml [hunk 8]
```diff
@@ -79,9 +79,32 @@ Resources:
...
+                              parts = bucket_arn.split(":::")[-1].split("/", 1)
+                              bucket_name = parts[0]
+                           
+                              # get the existing notification configuration
+                              existing_config = s3.get_bucket_notification_configuration(Bucket=bucket_name)
+
+                              existing_lambda_configs = existing_config.get('LambdaFunctionConfigurations', [])
+                              updated_lambda_configs = [config for config in existing_lambda_configs if config['LambdaFunctionArn'] != lambda_arn]
+ 
+                              notification_config = {
```
Comment (maya-jha) [inline s3-trigger-stack.yaml:97]: Do we need to check prefix also?

**Review by maya-jha — CHANGES_REQUESTED** [review]

File: lambda-cloudwatch-trigger-stack.yaml (hunk 1 above)
Comment (hrai-nr) [reply lambda-cloudwatch-trigger-stack.yaml:121]: Changed it to **NewRelicLogsLambdaSubscription_** as per naming convention for other resources.

File: s3-trigger-stack.yaml (hunk 7 above)
Comment (hrai-nr) [reply s3-trigger-stack.yaml:86]: Done

File: s3-trigger-stack.yaml (hunk 6 above)
Comment (hrai-nr) [reply s3-trigger-stack.yaml:181]: it was same code in both if and else, so removed from both and keep it outside of if and else

File: s3-trigger-stack.yaml (hunk 4 above)
Comment (hrai-nr) [reply s3-trigger-stack.yaml:107]: Removed.

File: s3-trigger-stack.yaml (hunk 3 above)
Comment (hrai-nr) [reply s3-trigger-stack.yaml:95]: Added comment # Deleting the existing configuration

File: s3-trigger-stack.yaml (hunk 2 above)
Comment (hrai-nr) [reply s3-trigger-stack.yaml:94]: Added the logic to continue.

File: s3-trigger-stack.yaml (hunk 8 above)
Comment (hrai-nr) [reply s3-trigger-stack.yaml:100]: It is working for prefix as well. Here is how I tested:
1. Create a bucket B1. and then create two folders (prefix) inside P1 and P2.
2. Go and manually add trigger for P1 and point it to some lambda xyz.
3. Now create a stack and give bucket B1 and prefix P2.
4. You can see two triggers for the bucket.
5. Delete the stack, only the trigger for P2 will be deleted since this will be associated with our lambda.

Let me know, if you want some other test.

**Review by maya-jha — APPROVED** [review]
